### PR TITLE
Sample limiter

### PR DESCRIPTION
**Pull request by parv3213**
This is a draft for a limiter contract. 
 There are two ways to go around this:
 1. Deploy a new custom limiter contract for each quest. (shown in this draft PR)
 2. Deploy multiple libraries with different limiter features. Set the specific limiter address while creating Quest (or even later). And pass in particular details like minimum token address or minimum token amount while making calls to the limiter.

### CODE REVIEW

**Comment by dan13ram** [comment]
Looks great. I was thinking exactly along these lines 🚀

**Comment by parv3213** [comment]
@dan13ram, I have a few concerns:
1. Should we allow users to create a custom limiter? Looking at this from a security point of view. 
2. If we have a particular use case for the limiters, then it may not be optimized to use these external limiters. For example, having a different set of QuestChain clones would be better/cheaper to provide the limiter checks.

**Comment by dan13ram** [comment]
> @dan13ram, I have a few concerns:
> 
> 1. Should we allow users to create a custom limiter? Looking at this from a security point of view.
> 2. If we have a particular use case for the limiters, then it may not be optimized to use these external limiters. For example, having a different set of QuestChain clones would be better/cheaper to provide the limiter checks.

I think you may be right. I was wondering if we can deploy some basic limiters such as the ones you've already written up.. 

But we can add a mapping of (questChainAddresss => tokenInfo ) for each of these limiters. And then we can let the quest chain owners set it up as they want it.. 

What do you think @parv3213  ?

**Comment by parv3213** [comment]
Hey, I do not understand what you mean by the `mapping of (questChainAddresss => tokenInfo ) ` bit. 

Maybe we can have a mapping where a limiter type (let's say token gating enum) will map to a `questChainTemplate` address. And inside the custom QuestChain clone, we can ask for a custom token address (for token gating).

**Comment by parv3213** [comment]
@dan13ram https://github.com/quest-chains/contracts/pull/17/commits/10324ef6a57b52e97ac2f1340af91695ee090981 contains the changes we discussed. I will be adding tests now. 

I haven't added supported for multi-token lib as it does not seems to be audited.

**Comment by parv3213** [comment]
Gas Report:

Without MultiTokenLib:
```
·-----------------------------------------------------|---------------------------|-----------------|-----------------------------·
|                Solc version: 0.8.16                 ·  Optimizer enabled: true  ·  Runs: 1000000  ·  Block limit: 30000000 gas  │
······················································|···························|·················|······························
|  Methods                                                                                                                        │
······················|·······························|·············|·············|·················|···············|··············
|  Contract           ·  Method                       ·  Min        ·  Max        ·  Avg            ·  # calls      ·  usd (avg)  │
······················|·······························|·············|·············|·················|···············|··············
|  LimiterTokenFee    ·  addQuestChainDetails         ·          -  ·          -  ·          99847  ·            3  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  LimiterTokenGated  ·  addQuestChainDetails         ·          -  ·          -  ·          76934  ·            3  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  MockERC20Token     ·  mint                         ·      34073  ·      68273  ·          51173  ·            4  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChain         ·  burnToken                    ·          -  ·          -  ·          37660  ·            3  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChain         ·  createQuests                 ·      35541  ·      52641  ·          44966  ·           13  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChain         ·  edit                         ·          -  ·          -  ·          29224  ·            5  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChain         ·  editQuests                   ·      34278  ·      36719  ·          35499  ·            6  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChain         ·  grantRole                    ·      48110  ·     136914  ·          96449  ·           30  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChain         ·  mintToken                    ·          -  ·          -  ·          66567  ·            9  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChain         ·  pause                        ·          -  ·          -  ·          32581  ·            3  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChain         ·  pauseQuests                  ·      59523  ·      79569  ·          69546  ·            6  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChain         ·  reviewProofs                 ·      43965  ·      53698  ·          45184  ·           25  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChain         ·  revokeRole                   ·      31803  ·      55065  ·          45092  ·           18  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChain         ·  setLimiter                   ·          -  ·          -  ·          52118  ·            6  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChain         ·  setTokenURI                  ·          -  ·          -  ·          50096  ·            3  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChain         ·  submitProofs                 ·      41097  ·      88130  ·          62956  ·           37  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChain         ·  unpause                      ·          -  ·          -  ·          32569  ·            3  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChainFactory  ·  create                       ·     404852  ·     739022  ·         449966  ·           44  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChainFactory  ·  createAndUpgrade             ·          -  ·          -  ·         457673  ·            3  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChainFactory  ·  createAndUpgradeWithPermit   ·          -  ·          -  ·         477626  ·            4  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChainFactory  ·  executeAdminReplace          ·          -  ·          -  ·          30433  ·            2  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChainFactory  ·  executePaymentTokenReplace   ·          -  ·          -  ·          31997  ·            2  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChainFactory  ·  executeUpgradeFeeReplace     ·          -  ·          -  ·          31856  ·            2  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChainFactory  ·  proposeAdminReplace          ·          -  ·          -  ·          69427  ·            2  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChainFactory  ·  proposePaymentTokenReplace   ·          -  ·          -  ·          69437  ·            2  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChainFactory  ·  proposeUpgradeFeeReplace     ·          -  ·          -  ·          68928  ·            2  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChainFactory  ·  upgradeQuestChain            ·          -  ·          -  ·          52360  ·            7  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChainFactory  ·  upgradeQuestChainWithPermit  ·          -  ·          -  ·          91948  ·            3  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  Deployments                                        ·                                             ·  % of limit   ·             │
······················································|·············|·············|·················|···············|··············
|  LimiterTokenFee                                    ·          -  ·          -  ·         315675  ·        1.1 %  ·          -  │
······················································|·············|·············|·················|···············|··············
|  LimiterTokenGated                                  ·          -  ·          -  ·         302907  ·          1 %  ·          -  │
······················································|·············|·············|·················|···············|··············
|  MockERC20Token                                     ·          -  ·          -  ·        1259349  ·        4.2 %  ·          -  │
······················································|·············|·············|·················|···············|··············
|  QuestChain                                         ·          -  ·          -  ·        3189001  ·       10.6 %  ·          -  │
······················································|·············|·············|·················|···············|··············
|  QuestChainFactory                                  ·          -  ·          -  ·        4217164  ·       14.1 %  ·          -  │
·-----------------------------------------------------|-------------|-------------|-----------------|---------------|-------------·
```

With MultiTokenLib:

```
·-----------------------------------------------------|---------------------------|-----------------|-----------------------------·
|                Solc version: 0.8.16                 ·  Optimizer enabled: true  ·  Runs: 1000000  ·  Block limit: 30000000 gas  │
······················································|···························|·················|······························
|  Methods                                                                                                                        │
······················|·······························|·············|·············|·················|···············|··············
|  Contract           ·  Method                       ·  Min        ·  Max        ·  Avg            ·  # calls      ·  usd (avg)  │
······················|·······························|·············|·············|·················|···············|··············
|  LimiterTokenFee    ·  addQuestChainDetails         ·          -  ·          -  ·          99847  ·            3  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  LimiterTokenGated  ·  addQuestChainDetails         ·          -  ·          -  ·          76934  ·            3  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  MockERC20Token     ·  mint                         ·      34073  ·      68273  ·          51173  ·            4  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChain         ·  burnToken                    ·          -  ·          -  ·          37660  ·            3  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChain         ·  createQuests                 ·      35541  ·      52641  ·          44966  ·           13  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChain         ·  edit                         ·          -  ·          -  ·          29224  ·            5  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChain         ·  editQuests                   ·      34278  ·      36719  ·          35499  ·            6  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChain         ·  grantRole                    ·      48110  ·     136914  ·          96449  ·           30  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChain         ·  mintToken                    ·          -  ·          -  ·          66567  ·            9  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChain         ·  pause                        ·          -  ·          -  ·          32581  ·            3  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChain         ·  pauseQuests                  ·      59523  ·      79569  ·          69546  ·            6  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChain         ·  reviewProofs                 ·      43965  ·      53698  ·          45184  ·           25  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChain         ·  revokeRole                   ·      31803  ·      55065  ·          45092  ·           18  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChain         ·  setLimiter                   ·          -  ·          -  ·          52118  ·            6  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChain         ·  setTokenURI                  ·          -  ·          -  ·          50096  ·            3  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChain         ·  submitProofs                 ·      41097  ·      88130  ·          62956  ·           37  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChain         ·  unpause                      ·          -  ·          -  ·          32569  ·            3  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChainFactory  ·  create                       ·     404852  ·     739022  ·         449966  ·           44  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChainFactory  ·  createAndUpgrade             ·          -  ·          -  ·         457673  ·            3  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChainFactory  ·  createAndUpgradeWithPermit   ·          -  ·          -  ·         477626  ·            4  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChainFactory  ·  executeAdminReplace          ·          -  ·          -  ·          30433  ·            2  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChainFactory  ·  executePaymentTokenReplace   ·          -  ·          -  ·          31997  ·            2  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChainFactory  ·  executeUpgradeFeeReplace     ·          -  ·          -  ·          31856  ·            2  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChainFactory  ·  proposeAdminReplace          ·          -  ·          -  ·          69427  ·            2  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChainFactory  ·  proposePaymentTokenReplace   ·          -  ·          -  ·          69437  ·            2  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChainFactory  ·  proposeUpgradeFeeReplace     ·          -  ·          -  ·          68928  ·            2  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChainFactory  ·  upgradeQuestChain            ·          -  ·          -  ·          52360  ·            7  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  QuestChainFactory  ·  upgradeQuestChainWithPermit  ·          -  ·          -  ·          91948  ·            3  ·          -  │
······················|·······························|·············|·············|·················|···············|··············
|  Deployments                                        ·                                             ·  % of limit   ·             │
······················································|·············|·············|·················|···············|··············
|  LimiterTokenFee                                    ·          -  ·          -  ·         315675  ·        1.1 %  ·          -  │
······················································|·············|·············|·················|···············|··············
|  LimiterTokenGated                                  ·          -  ·          -  ·         302907  ·          1 %  ·          -  │
······················································|·············|·············|·················|···············|··············
|  MockERC20Token                                     ·          -  ·          -  ·        1259349  ·        4.2 %  ·          -  │
······················································|·············|·············|·················|···············|··············
|  QuestChain                                         ·          -  ·          -  ·        3189001  ·       10.6 %  ·          -  │
······················································|·············|·············|·················|···············|··············
|  QuestChainFactory                                  ·          -  ·          -  ·        4217164  ·       14.1 %  ·          -  │
·-----------------------------------------------------|-------------|-------------|-----------------|---------------|-------------·
```

**Review by dan13ram — COMMENTED** [review]
Looking really good man 🔥

**Comment by parv3213** [comment]
> Looking really good man 🔥

Thanks man. I still need to add the multi-token Library to the limiter token gated.

**Comment by parv3213** [comment]
> Looking really good man 🔥

Added MultiToken to both limiters. 
Note: The deployment cost of both limiters has almost doubled.

**Comment by dan13ram** [comment]
> Added MultiToken to both limiters. Note: The deployment cost of both limiters has almost doubled.

I think that's fine, the per call cost is not too high right?

**Comment by parv3213** [comment]
> > Added MultiToken to both limiters. Note: The deployment cost of both limiters has almost doubled.
> 
> I think that's fine, the per call cost is not too high right?

Call costs have not increased significantly.

File: contracts/LimiterTokenGated.sol [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        );
+        emit AddQuestChainDetails(
+            _questChain,
+            _tokenAddress,
+            _minBalance,
+            msg.sender
+        );
+    }
+
+    function submitProofLimiter(address _sender) external {
```
Comment (dan13ram) [inline contracts/LimiterTokenGated.sol]: think of a better name

**Comment by parv3213** [comment]
@dan13ram This limiter branch is ready for review.

File: contracts/QuestChain.sol [hunk 2]
```diff
@@ -13,9 +13,11 @@ import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
 import "@openzeppelin/contracts/utils/Context.sol";
 
 import "./interfaces/IQuestChain.sol";
+import "./interfaces/ILimiter.sol";
 
 // author: @dan13ram
```
Comment (dan13ram) [inline contracts/QuestChain.sol]: can remove this line and just keep the new line with both handles

**Review by dan13ram — APPROVED** [review]
Looks and works great. I added some comments and set a fixed solc version for MultiToken.sol

**Pull request merged**